### DISCUSSION TITLE
Background difference between transform() and trasformToUni()

### DISCUSSION THREAD

**Discussion by ichrist-gr** [quarkusio/quarkus#23481] [Q&A] [answered]
So mongoService.retrieveClothesList(clothesDto) is an async action. 

Noticed that onItem().transformToUni() is equivalent to flatmap()

```java
public Uni<List<Clothes>> retrieve(ClothesDto clothesDto) {
    logger.info("ClothesDto: " + clothesDto);
    return mongoService
        .retrieveClothesList(clothesDto)
        .onItem()
        .transformToUni(this::handleListResponse);
  }

  private Uni<List<Clothes>> handleListResponse(List<Clothes> clothes) {
    if (Objects.isNull(clothes)) {
      logger.info("Mongo failed to retrieve clothes list");
      return Uni.createFrom().nullItem();
    }

    logger.info(clothes.toString());
    return Uni.createFrom().item(clothes);
  }
```

Although, i have a question that will help me deeply understand the async meaning and the response handling of this action.

```java
public Uni<List<Clothes>> retrieve(ClothesDto clothesDto) {
    logger.info("ClothesDto: " + clothesDto);
    return mongoService
        .retrieveClothesList(clothesDto)
        .onItem()
        .transform(this::handleListResponse);
  }

  private List<Clothes> handleListResponse(List<Clothes> clothes) {
    if (Objects.isNull(clothes)) {
      logger.info("Mongo failed to retrieve clothes list");
      return null;
    }

    logger.info(clothes.toString());
    return clothes;
  }
```

The output from mongo service has calculated because of uni, but is it necessary to pass the output and convert the whole method to Uni?
So, i need a clarification about what these response handlers do in background and which is the right async way. 

Thank you in advance

**Comment by jponge** [comment, marked as answer]
I didn't get what was your exact question, but I'll explain here the difference between the 2 methods.

`transform` does a _synchronous_ transformation, and you should use it for anything that does not have I/O involved.

`transformToUni` should be used when you want to chain the processing with an operation that does some async I/O.

For instance if operation `A` gives you an object `foo`:

- use `transform` if you do something like `foo -> foo.getName()`
- use `transformToUni` if you do something like `foo -> db.insert(foo)`.

See https://smallrye.io/smallrye-mutiny/getting-started/transforming-items and https://smallrye.io/smallrye-mutiny/getting-started/transforming-items-async

**Reply by ichrist-gr** [reply]
That's what i had in mind. So in my case,list is returned after validation, so there is no need of transformToUni(). 

Thank you very much
